### PR TITLE
Normalize Object Key when Getting Resources

### PR DESCRIPTION
**Pull request by michaelahlers**
Possible solution for ohnosequences/sbt-s3-resolver#52.

### CODE REVIEW

File: src/main/java/ohnosequences/ivy/S3Repository.java [hunk 1]
```diff
@@ -117,6 +117,8 @@ public void get(String source, File destination) {
   }
 
   public Resource getResource(String source) {
+    /* For ohnosequences/sbt-s3-resolver#52, has the effect of cleaning up redundant path delimiters (thereby fixing invalid S3 object keys). */
+    source = java.net.URI.create(source).normalize().toString();
```
Comment (laughedelic) [inline src/main/java/ohnosequences/ivy/S3Repository.java]: Take a look at the `S3Utils.getUri` method:

https://github.com/michaelahlers/ivy-s3-resolver/blob/df2ffd3dc5010139b9ac0276646dc301dd1a7bd5/src/main/java/ohnosequences/ivy/S3Utils.java#L50-L56

I think you can add there `.normalize` and then this change will propagate to the `getBucket`, `getKey` methods used in `S3Resolver.put`. 
And then here in `getResource` you can also use it as `S3Utils.getUri(source).toString()`.

**Comment by michaelahlers** [comment]
@laughedelic, found a little time today, so I'll try wrapping this up!

**Comment by michaelahlers** [comment]
@laughedelic, any objection to adding a (Scala-fluent) unit test framework to this project?

**Comment by laughedelic** [comment]
@michaelahlers of course, you are more than welcome to add tests here! 

This library was maintained by my colleague before, and since it was handed to me, I was keeping the changes to minimum. I was actually thinking recently to drop this dependency in sbt-s3-resolver and just write an analogous resolver in Scala. But I don't know yet when I will have time for doing it...

File: src/test/scala/ohnosequences/ivy/S3UtilsSpec.scala [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * @author <a href="michael@ahlers.consulting">Michael Ahlers</a>
+ */
+class S3UtilsSpec extends WordSpec with Matchers {
+
+  "Getting buckets and keys" when {
+
+    val scenarios =
+      ("s3://bucket/p0/p1", ("bucket", "p0/p1")) ::
+        ("s3://bucket/p0//p1", ("bucket", "p0/p1")) ::
+        Nil
```
Comment (michaelahlers) [inline src/test/scala/ohnosequences/ivy/S3UtilsSpec.scala]: For now, only includes basic and ohnosequences/sbt-s3-resolver#52 cases.

File: build.sbt [hunk 3]
```diff
@@ -15,6 +15,21 @@ libraryDependencies ++= Seq(
...
+  Wart.DefaultArguments,
+  Wart.NonUnitStatements,
+  Wart.Null
+)
+
+libraryDependencies ++= Seq(
+  "com.github.pathikrit" %% "better-files" % "2.17.1" % Test,
+  "org.scalamock" %% "scalamock-scalatest-support" % "3.6.0" % Test,
+  "org.scalatest" %% "scalatest" % "3.0.1" % Test
+)
```
Comment (michaelahlers) [inline build.sbt]: Add test libraries with helpful support, and make Wartremover less strict from `"ohnosequences" % "nice-sbt-settings" % "0.8.0"` settings.

File: src/main/java/ohnosequences/ivy/S3Repository.java [hunk 4]
```diff
@@ -88,10 +88,29 @@ public S3Repository(
...
+   * Package-private constructor specifically for taking an {@link AmazonS3} instance that can be mocked under test.
+   */
+  S3Repository(
+    AmazonS3 s3Client,
+    boolean overwrite,
+    CannedAccessControlList acl,
+    boolean serverSideEncryption,
+    StorageClass storageClass
+  ) {
+    this.s3Client = s3Client;
```
Comment (michaelahlers) [inline src/main/java/ohnosequences/ivy/S3Repository.java]: An unfortunate conceit. This is unrelated to the fix, but it makes life significantly easier to support injection of (mocked) `AmazonS3` instances when `S3Repository` is under test.

File: src/main/java/ohnosequences/ivy/S3Utils.java [hunk 5]
```diff
@@ -49,7 +49,8 @@ public static String getKey(String uri) {
 
   private static URI getUri(String uri) {
     try {
-      return new URI(uri);
+      /* For ohnosequences/sbt-s3-resolver#52, has the effect of cleaning up redundant path delimiters (thereby fixing invalid S3 object keys). */
+      return new URI(uri).normalize();
```
Comment (michaelahlers) [inline src/main/java/ohnosequences/ivy/S3Utils.java]: Crux of the fix. Everything ultimately comes through here.

File: src/test/scala/ohnosequences/ivy/S3MockableRepository.scala [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  acl: CannedAccessControlList = CannedAccessControlList.PublicRead,
+  serverSideEncryption: Boolean = false,
+  storageClass: StorageClass = StorageClass.Standard
+) extends S3Repository(
+  s3Client,
+  overwrite,
+  acl,
+  serverSideEncryption,
+  storageClass
+)
```
Comment (michaelahlers) [inline src/test/scala/ohnosequences/ivy/S3MockableRepository.scala]: Help make tests more concise where all they care about is the client.

File: src/test/scala/ohnosequences/ivy/S3RepositorySpec.scala [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            (client.putObject(_: PutObjectRequest)).expects(where(requestPredicate))
+
+            val repository = new S3MockableRepository(client)
+            repository.put(source.toJava, given, true)
+          }
+        }
+    }
+  }
+
+}
```
Comment (michaelahlers) [inline src/test/scala/ohnosequences/ivy/S3RepositorySpec.scala]: Test cases for each API mentioned in the review.

File: src/test/scala/ohnosequences/ivy/S3ResourceSpec.scala [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+              'getLastModified (lastModified),
+              'getName (given)
+            )
+          }
+        }
+    }
+
+  }
+
+}
```
Comment (michaelahlers) [inline src/test/scala/ohnosequences/ivy/S3ResourceSpec.scala]: This is used as a delegate by `S3Repository.get`, so ensure it's behavior is correct.

File: src/test/scala/ohnosequences/ivy/S3UtilsSpec.scala [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+          }
+          s"""return object key "$objectKey"""" in {
+            S3Utils.getKey(given) should equal(objectKey)
+          }
+        }
+    }
+
+  }
+
+}
```
Comment (michaelahlers) [inline src/test/scala/ohnosequences/ivy/S3UtilsSpec.scala]: Redundant, yes, but it'd take significant refactoring to remove various assumptions about URI parsing.

File: src/test/scala/ohnosequences/ivy/Scenarios.scala [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  val locationWithBucketNameAndObjectKeys: List[(String, (String, String))] =
+    ("s3://bucket/path0", ("bucket", "path0")) ::
+      ("s3://bucket//path0", ("bucket", "path0")) ::
+      ("s3://bucket/path0/path1", ("bucket", "path0/path1")) ::
+      ("s3://bucket/path0//path1", ("bucket", "path0/path1")) ::
+      ("s3://bucket//path0/path1", ("bucket", "path0/path1")) ::
+      ("s3://bucket//path0//path1", ("bucket", "path0/path1")) ::
+      Nil
+
+}
```
Comment (michaelahlers) [inline src/test/scala/ohnosequences/ivy/Scenarios.scala]: Not my favorite, but hard-code scenarios we want to address with this pull-request.

**Comment by michaelahlers** [comment]
@laughedelic, ready for another review. There's a lot added to `src/test/scala` here, but I figured aggressively adding test coverage to essential functionality couldn't hurt while covering the specific bug. A lot of this test code could be removed with the introduction of a type representing normalized and validated S3 object identifiers, but that may be done in another PR (or not at all depending on whether or not this library's scrapped).

**Comment by laughedelic** [comment]
@michaelahlers wow! 💯  I didn't expect that you will write tests for other parts besides `getKey`/`getBucket`. This is such a great contribution! I'm going to review the code tomorrow as soon as a have time.

File: src/test/scala/ohnosequences/ivy/S3RepositorySpec.scala [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+              val listing = mock[ObjectListing]
+              (listing.getCommonPrefixes _).expects().once().returns(directories)
+              (listing.getObjectSummaries _).expects().once() returns {
+                files map { file =>
+                  val summary = mock[S3ObjectSummary]
+                  (summary.getKey _).expects().once().returns(file)
+                  summary
+                }
+              }
+              (listing.getNextMarker _).expects().once().returns(null)
```
[the remark this answers is not in the document]
Comment (laughedelic) [reply src/test/scala/ohnosequences/ivy/S3RepositorySpec.scala]: No problem with it. I told it to ignore this particular place.

**Comment by laughedelic** [comment]
Released in [v0.12.0](https://github.com/ohnosequences/ivy-s3-resolver/releases/tag/v0.12.0) :shipit: 

I also added you to the collaborators. In case you'll want to hack around here some more, you're welcome to submit PRs directly 😉

**Comment by michaelahlers** [comment]
@laughedelic, what can I say? I love writing test code. 😉 You're welcome, I'm glad to help, and thanks for the invitation. (Accepted!) Your S3 resolver plugin has been extremely useful to me on a number of projects both professional and personal. Pretty thrilled to contribute back any way I can!

**Pull request merged**